src/smc-webapp/project_store.coffee

```diff
@@ -1200,18 +1210,29 @@ class ProjectActions extends Actions
                 if target[target.length-1] == '/' or full_path == ''
                     @open_directory(parent_path)
                 else
-                    @fetch_directory_listing(parent_path, true, true)
-                    @get_store().wait
-                        until   : (s) =>
-                            listing = s.directory_listings.get(parent_path)
-                            return listing?.find (val) => val.get('name') == last
-                        timeout : 30
-                        cb      : (err, item) =>
-                            if err
-                                alert_message(type:'error', message:"There was an error related to opening the link: #{err}")
-                                @open_directory(parent_path)
-                            else
-                                if item.get('isdir')
+                    # TODOJ: Change when directory listing is synchronized. Just have to query client state then.
+                    # Assume that if it's loaded, it's good enough.
+                    store = @get_store()
+                    listing = store.directory_listings.get(parent_path)
+                    item = listing?.find (val) => val.get('name') == last
+                    if item?
+                        if item.get('isdir')
+                            @open_directory(full_path)
+                        else
+                            @open_file
+                                path       : full_path
+                                foreground : foreground
+                                foreground_project : foreground
+                    else
+                        @fetch_directory_listing
+                            path         : parent_path
+                            show_hidden  : true
+                            finish_cb    : =>
+                                store = @get_store()
+                                listing = store.directory_listings.get(parent_path)
+                                item = listing?.find (val) => val.get('name') == last
```

Comment:
Want to try exactly once after `fetch_directory_listing` so `store.wait` isn't any good.